diff -uNr dynamips.v2/common/device.h dynamips/common/device.h
--- dynamips.v2/common/device.h	2011-01-23 18:45:21.000000000 +0000
+++ dynamips/common/device.h	2011-03-21 13:13:15.225610047 +0000
@@ -45,6 +45,35 @@
 #include "pci_dev.h"
 
 /* device access function */
+#ifdef MAC64HACK
+static void *__dev_access_fast(cpu_gen_t *cpu,u_int dev_id,m_uint32_t offset,
+                      u_int op_size,u_int op_type,m_uint64_t *data)
+{
+   struct vdevice *dev = cpu->vm->dev_array[dev_id];
+
+   if (unlikely(!dev)) {
+      cpu_log(cpu,"dev_access_fast","null handler (dev_id=%u,offset=0x%x)\n",
+              dev_id,offset);
+      return NULL;
+   }
+
+#if DEBUG_DEV_PERF_CNT
+   cpu->dev_access_counter++;
+#endif
+
+   return(dev->handler(cpu,dev,offset,op_size,op_type,data));
+}
+
+static void *dev_access_fast(cpu_gen_t *cpu,u_int dev_id,m_uint32_t offset,
+                    u_int op_size,u_int op_type,m_uint64_t *data)
+{
+  asm("sub $8, %rsp");
+  void* ret = __dev_access_fast(cpu, dev_id, offset, op_size, op_type, data);
+  asm("add $8, %rsp");
+  return ret;
+}
+
+#else
 static forced_inline 
 void *dev_access_fast(cpu_gen_t *cpu,u_int dev_id,m_uint32_t offset,
                       u_int op_size,u_int op_type,m_uint64_t *data)
@@ -63,6 +92,7 @@
 
    return(dev->handler(cpu,dev,offset,op_size,op_type,data));
 }
+#endif
 
 /* Get device by ID */
 struct vdevice *dev_get_by_id(vm_instance_t *vm,u_int dev_id);
diff -uNr dynamips.v2/Makefile dynamips/Makefile
--- dynamips.v2/Makefile	2011-03-21 10:56:14.165484991 +0000
+++ dynamips/Makefile	2011-03-21 13:25:13.513485071 +0000
@@ -11,8 +11,9 @@
 #export DYNAMIPS_LIB?=lib64
 export DYNAMIPS_LIB?=lib
 
-# For MAC x64 you need to compile the "unstable" version
-# Otherwise you should really choose "stable" here.
+# For MAC x64 you can compile the "unstable" version, which should work
+# fine, or use stable if you prefer.
+# For all other targets you should really choose "stable" here.
 # Unstable has some optimizations in it, plus new TLA code.
 # However it does not seem stable for non-mac users
 export DYNAMIPS_CODE?=stable
diff -uNr dynamips.v2/stable/Makefile dynamips/stable/Makefile
--- dynamips.v2/stable/Makefile	2011-03-21 10:50:39.000000000 +0000
+++ dynamips/stable/Makefile	2011-03-21 13:24:37.953609993 +0000
@@ -83,6 +83,9 @@
    OSNAME=Darwin
    CFLAGS+=-I/usr/local/include -I/opt/local/include -mdynamic-no-pic -D_FILE_OFFSET_BITS=64
    LIBS=-L/usr/local/lib -L/opt/local/lib -L. -ldl -lelf -lpthread
+ifeq ($(DYNAMIPS_ARCH),amd64)
+   CFLAGS+= -DMAC64HACK
+endif
 else
 ifeq ($(shell uname -s), SunOS)
    OSNAME=SunOS

